src/libraries/Native/Unix/System.Native/pal_networking.c

```diff
@@ -432,6 +422,137 @@ int32_t SystemNative_GetHostEntryForName(const uint8_t* address, HostEntry* entr
     return ret;
 }
 
+#if HAVE_GETADDRINFO_A
+struct GetAddrInfoAsyncState
+{
+    struct gaicb gai_request;
+    struct gaicb* gai_requests;
+    struct sigevent sigevent;
+
+    const uint8_t* address;
+    HostEntry* entry;
+    GetHostEntryForNameCallback callback;
+};
+
+static void GetHostEntryForNameAsyncComplete(sigval_t context)
+{
+    struct GetAddrInfoAsyncState* state = (struct GetAddrInfoAsyncState*)context.sival_ptr;
+
+    atomic_thread_fence(memory_order_acquire);
+
+    GetHostEntryForNameCallback callback = state->callback;
+
+    int result = gai_error(&state->gai_request);
+    HostEntry* entry = state->entry;
+
+    int ret = ConvertGetAddrInfoAndGetNameInfoErrorsToPal(result);
+
+    if (result == 0)
+    {
+        const uint8_t* address = state->address;
+        struct addrinfo* info = state->gai_request.ar_result;
+
+        ret = GetHostEntries(address, info, entry);
+    }
+
+    free(state);
+
+    if (callback != NULL)
+    {
+        callback(entry, ret);
+    }
```

Comment:
```suggestion
    int ret = ConvertGetAddrInfoAndGetNameInfoErrorsToPal(gai_error(&state->gai_request));

    if (ret == 0)
    {
        const uint8_t* address = state->address;
        struct addrinfo* info = state->gai_request.ar_result;

        ret = GetHostEntries(address, info, entry);
    }

    if (callback != NULL)
    {
        callback(state->entry, ret);
    }

    free(state);
```

Reply:
I had the `free(state)` intentionally before calling the callback.

* `entry` is passed from managed to native and via callback back to managed -- so the position of `free(state)` doesn't change anything
* if the callback (managed side) has an exception `state` won't be freed an so this can be a memory leak -- although `ProcessResult` on the managed side (the "callback worker") has a `try-finally`

I prefer leaving the `free(state)` before the callback. When there is a good reason to move it after the callback, I'll do it for sure.

Reply:
Both are equivalent, yes, but the suggested change is slightly shorter (and doesn't have a variable that's attributed to but never read from, which is cleaner).

Reply:
OK.

Reply:
(It's just a suggestion, no need to change it at all.)

Reply:
Yeah, but I like it as the code looks cleaner.
Anyway I highly appreciate the feedback here! (My C is a bit rusty, so it's good to get some hints.)